File: Collatz Conjecture/collatz_conjecture.py
'''
Each number is obtained from the previous iteration.
Start with x
If x is odd 3x+1
If x is even x/2
Repeat

So far all numbers tested end with a 4,2,1 loop.

Will all positive integers eventually reach 1?
Do all result in the same 4,2,1 loop
'''

import sys, math

def collatz(x):
    # Do collatz stuff
    x = int(x)
    result = []

    print("collatz: " + str(x))
    while(x >= 1):
        result.append(math.trunc(x))
        # Abort if 1 has been reached
        if x == 1:
            break

        if x % 2 == 0:
            x = x / 2
        else:
            x = (x * 3) + 1

    return result


def main(args):
    results = "ERROR: No result"
    if len(args) >= 1:
        if args[0].isdigit():
            results = collatz(args[0])
        else:
            print("Please supply a number like this: \n" + sys.argv[0] + " 12")
    else:
        startNum = input("Please enter a starting number: ")
        if startNum.isdigit():
            results = collatz(startNum)
    
    print(results)


if __name__ == "__main__":
    # execute only if run as a script
    main(sys.argv[1:])

File: Collatz Conjecture/Lua/collatz_conjecture.lua
#!/usr/bin/env lua
--[[
--Each number is obtained from the previous iteration.
--Start with x
--If x is odd 3x+1
--If x is even x/2
--Repeat until you reach 1
--
--The question trying to be solved is:
--Will all positive integers eventually reach 1?
--Do all numbers result in the same 4, 2, 1 loop?
]]--

local start_num
local good = "no"
local collatz = ""
local current_value = 0

repeat
    io.write("Please input a starting number: ")
    start_num = io.read()
    start_num = tonumber(start_num)
    if type(start_num) == "number" then
        good = "yes"
    else
        print("Input was not a number")
    end
until good == "yes"

current_value = start_num
repeat 
    if current_value % 2 == 0 then
        -- current is even
        current_value = current_value / 2
    else
        -- current is odd
        current_value = (3 * current_value) + 1
    end
    collatz = collatz .. current_value .. " "
until current_value == 1

print("num: ", start_num)
print("collatz: ", collatz)
